runtime/tests/checker/range_value_test.go

```diff
@@ -377,3 +377,48 @@ func TestCheckInclusiveRangeConstructionInvalid(t *testing.T) {
 		[]error{&sema.MissingTypeArgumentError{}},
 	)
 }
+
+func TestInclusiveRangeNonLeafIntegerTypes(t *testing.T) {
+
+	t.Parallel()
+
+	baseValueActivation := sema.NewVariableActivation(sema.BaseValueActivation)
+	baseValueActivation.DeclareValue(stdlib.InclusiveRangeConstructorFunction)
+
+	options := ParseAndCheckOptions{
+		Config: &sema.Config{
+			BaseValueActivation: baseValueActivation,
+		},
+	}
+
+	test := func(t *testing.T, ty sema.Type) {
+		t.Run(fmt.Sprintf("InclusiveRange<%s>", ty), func(t *testing.T) {
+			t.Parallel()
+
+			_, err := ParseAndCheckWithOptions(t, fmt.Sprintf(`
+				let a: %[1]s = 0
+				let b: %[1]s = 10
+				var range: InclusiveRange<%[1]s> = InclusiveRange<%[1]s>(a, b)
```

Comment:
Yes, I missed that point. Without further changes, the expression `var range = InclusiveRange<Integer>(a, b)` didn't return any checker error since the RHS is a function invocation. The name `InclusiveRange` is used in both the type as well as the constructor function.

I had to implement the same check in the constructor function as well. This required me to implement support for custom type parameter validation in `FunctionType`. I did that as part of 7b65121e7.

Reply:
Very nice @darkdrag00nv2!